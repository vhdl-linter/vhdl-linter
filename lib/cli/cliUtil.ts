import { readdirSync } from "fs";
import { cwd } from "process";
import { DiagnosticSeverity } from "vscode-languageserver";
import { OIRange } from "../parser/objects";
import { joinURL } from "../projectParser";
import { OIDiagnostic } from "../vhdlLinter";

export interface MessageWrapper {
  file: string,
  messages: (OIDiagnostic | { message: string })[]
}
function isOIDiagnostic(obj: unknown): obj is OIDiagnostic {
  if ((obj as OIDiagnostic).range instanceof OIRange) {
    return true;
  }
  return false;
}
export function readDirPath(path: URL) {
  return readdirSync(path).map(file => joinURL(path, file));
}
function getMessageColor(message: OIDiagnostic | { message: string }) {
  if (isOIDiagnostic(message) && message.severity === DiagnosticSeverity.Error) {
    return '\u001b[31m';
  } else if (isOIDiagnostic(message) && message.severity === DiagnosticSeverity.Warning) {
    return '\u001b[33m';
  }
  return '\u001b[34m';
}
export function prettyPrintMessages(messages: MessageWrapper[]) {
  return messages.map(message => {
    const filename = message.file.replace(cwd(), '');
    return message.messages.slice(0, 5).map((innerMessage) => {
      const messageText = `${getMessageColor(innerMessage)}${innerMessage.message}\u001b[0m`;
      if (isOIDiagnostic(innerMessage)) {
        return `${filename}:${innerMessage.range.start.line + 1} (r: ${innerMessage.range.start.line}:${innerMessage.range.start.character} - ${innerMessage.range.end.line}:${innerMessage.range.end.character})\n  ${messageText}`; // lines are 0 based in OI
      }
      return `${filename}\n  ${messageText}`;
    }).join('\n') + (message.messages.length > 5 ? `\n\u001b[31m ... and ${message.messages.length - 5} more\u001b[0m` : '');
  }).join('\n');
}
