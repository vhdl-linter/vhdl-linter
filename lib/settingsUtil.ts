import { DeepPartial } from "utility-types";
import { ISettings } from "./settingsGenerated";
import { existsSync, readFileSync } from "fs";
import { join } from "path";


// these are capabilities from the language server and mainly used by the language server
// However, the project parser might use the `configuration` to determine whether to get settings from the language server.
// We define them here and not in the languageServer to not start one from the projectParser
export const currentCapabilities = {
  configuration: false,
  workspaceFolder: false
};

export function normalizeSettings(settings: ISettings) {
  const newSettings = JSON.parse(JSON.stringify(settings)) as ISettings;
  for (const [key, value] of Object.entries(newSettings.style)) {
    if (typeof (newSettings.style as Record<string, unknown>)[key] === 'string') {
      (newSettings.style as Record<string, unknown>)[key] = (value as string).trim();
    }
  }
  return newSettings;
}

export function overwriteSettings(settings: ISettings, overwrite: DeepPartial<ISettings>) {
  const newSettings = JSON.parse(JSON.stringify(settings)) as ISettings;
  recursiveObjectAssign(newSettings, overwrite);
  return normalizeSettings(newSettings);
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function recursiveObjectAssign<T extends Record<string, any>>(target: T, source: DeepPartial<T>) {
  Object.keys(source).forEach(key => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const s_val = source[key];
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const t_val = target[key];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/strict-boolean-expressions
    (target as any)[key] = t_val && s_val && typeof t_val === 'object' && typeof s_val === 'object'
      ? recursiveObjectAssign(t_val, s_val)
      : s_val;
  });
  return target;
}
type primitive = string | boolean | number;
interface Schema {
  [key: string]: Schema | primitive[] | primitive;
}
// for initial transpile before doing ts-node generateSettings, the file does not exist...
const schemaFile = join(__dirname, '../settings.schema.json');
const file = existsSync(schemaFile) ? JSON.parse(readFileSync(join(__dirname, '../settings.schema.json'), { encoding: 'utf-8' })) as Schema : {};
function objectWalk(object: Schema) {
  const result: Schema = {};
  for (const [key, value] of Object.entries(object)) {
    if (key !== 'deprecationMessage') {
      if (typeof value === 'object' && !Array.isArray(value)) {
        result[key] = objectWalk(value);
      } else {
        result[key] = value;
      }
    }
  }
  return result;
}
export const settingsSchema = objectWalk(file);