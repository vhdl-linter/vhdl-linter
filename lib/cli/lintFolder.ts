import { lstatSync } from "fs";
import { ProjectParser, vhdlGlob } from "../projectParser";
import { VhdlLinter } from "../vhdlLinter";
import { readFileSyncNorm } from "./readFileSyncNorm";
import { MessageWrapper, prettyPrintMessages, readDirPath } from "./cliUtil";
import { OIRange } from "../parser/objects";
import { minimatch } from "minimatch";
import { basename, dirname } from "path";
import { fileURLToPath } from "url";
import { execSync } from "child_process";
import { platform } from "process";

let gitExists = false;
if (platform === 'win32') {
  try {
    execSync(`git --version`);
    gitExists = true;
  } catch (e) {
    gitExists = false;
  }
} else {
  try {
    execSync(`command -v git`);
    gitExists = true;
  } catch (e) {
    gitExists = false;
  }
}
function isGitIgnored(path: string) {
  if (path.endsWith('.git')) {
    return true;
  }
  if (gitExists) {
    // git check-ignore will fail if the file is *NOT* ignored or when we are not in a git repository
    try {
      execSync(`git check-ignore ${path}`, {cwd: dirname(path)});
      // is a git repo and the file is ignored...
      return true;
    } catch (e) {
      return false;
    }
  } else {
    return false;
  }
}
function ignore(excludePatterns: string[], relativePath: string, absolutePath: string) {
  return excludePatterns.some(glob => minimatch(relativePath, glob)) || isGitIgnored(absolutePath);
}

export async function lintFolder(path: URL, errorExpected: boolean, printMessages: boolean, excludePatterns: string[], projectParser: ProjectParser): Promise<MessageWrapper[]> {
  const result: MessageWrapper[] = [];
  const rootPath = fileURLToPath(projectParser.workspaces[0]!);
  for (const subPath of readDirPath(path)) {
    // in this case we only ever have one workspace
    const relativePath = fileURLToPath(subPath).replace(rootPath, '').substring(1);
    // ignore excluded folders & files
    if (lstatSync(subPath).isDirectory()) {
      if (ignore(excludePatterns, relativePath, fileURLToPath(subPath))) {
        continue;
      }
      result.push(...await lintFolder(subPath, errorExpected, printMessages, excludePatterns, projectParser));
    } else if (minimatch(basename(relativePath), vhdlGlob)) {
      if (ignore(excludePatterns, relativePath, fileURLToPath(subPath))) {
        continue;
      }
      const text = readFileSyncNorm(subPath, { encoding: 'utf8' });
      const vhdlLinter = new VhdlLinter(subPath, text, projectParser, await projectParser.getDocumentSettings(subPath));
      if (vhdlLinter.parsedSuccessfully) {
        await vhdlLinter.checkAll();
      }
      if (errorExpected === false) {
        if (vhdlLinter.messages.length > 0) {
          const newMessage = {
            file: fileURLToPath(subPath),
            messages: vhdlLinter.messages
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages(rootPath, [newMessage]));
          }
        }
      } else {
        if (vhdlLinter.messages.length !== 1) {
          const newMessage: MessageWrapper = {
            file: fileURLToPath(subPath),
            messages: [
              ...vhdlLinter.messages,
              {
                message: `One message expected found ${vhdlLinter.messages.length}`,
                range: new OIRange(vhdlLinter.file, 0, 0)
              }]
          };
          result.push(newMessage);
          if (printMessages) {
            console.log(prettyPrintMessages(rootPath, [newMessage]));
          }
        }
      }
    }
  }

  return result;
}