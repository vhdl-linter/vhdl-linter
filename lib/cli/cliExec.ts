import { Command, InvalidArgumentError } from '@commander-js/extra-typings';
import { existsSync, lstatSync } from 'fs';
import { isAbsolute, join } from 'path';
import { cwd } from 'process';
import { fileURLToPath, pathToFileURL } from 'url';
import { DiagnosticSeverity } from 'vscode-languageserver';
import { ProjectParser } from '../projectParser';
import { OIDiagnostic } from '../vhdlLinter';
import { getCodeClimate, prettyPrintMessages } from './cliUtil';
import { lintFolder } from './lintFolder';

export async function cli(argv: string[]) {
  const integerParam = (v: string) => {
    const n = Number(v);
    if (Number.isInteger(n)) {
      return n;
    }
    throw new InvalidArgumentError(`Not an integer`);
  };
  const folderParam = (v: string) => {
    const path = isAbsolute(v) ? v : join(cwd(), v);
    if (existsSync(path) === false) {
      throw new InvalidArgumentError("Does not exist");
    }
    if (lstatSync(path).isDirectory() === false) {
      throw new InvalidArgumentError("Is not a directory");
    }
    return pathToFileURL(path);
  };
  const program = new Command('vhdl-linter')
    .description('A typescript based linter for vhdl')
    .argument('<folder>', 'The folder to lint', folderParam)
    .option('-j, --output-json', 'Output message in json compatible with Code Climate Engine Specification (Gitlab & Github compatible)', false)
    .option('-e, --exclude <pattern...>', 'Exclude pattern for linting. Use the ignore setting from the `vhdl-linter.yml` file to not parse files at all.', [])
    .option('-i, --max-info <count>', 'Number of information messages to trigger a nonzero exit code', integerParam, -1)
    .option('-w, --max-warnings <count>', 'Number of information messages to trigger a nonzero exit code', integerParam, -1)
    .parse(argv);
  const options = program.opts();
  const start = new Date().getTime();
  const folder = program.processedArgs[0];
  const projectParser = await ProjectParser.create([folder]);
  const filesWithMessages = await lintFolder(folder, false, options.outputJson === false, options.exclude, projectParser);
  await projectParser.stop();
  const allMessages = filesWithMessages.flatMap(m => m.messages);
  const warningCount = allMessages.filter(m => m?.severity === DiagnosticSeverity.Warning).length;
  const infoCount = allMessages.filter(m => m?.severity === DiagnosticSeverity.Information).length;
  // default is error
  function isError(m: OIDiagnostic) {
    return m?.severity !== DiagnosticSeverity.Warning && m?.severity !== DiagnosticSeverity.Information;
  }
  const errorCount = allMessages.filter(isError).length;
  if (options.outputJson) {
    const rootPath = fileURLToPath(projectParser.workspaces[0]!);
    console.log(JSON.stringify(getCodeClimate(filesWithMessages, rootPath), undefined, 2));
    const filesWithErrors = filesWithMessages.map(fileWithMessages => ({ ...fileWithMessages, messages: fileWithMessages.messages.filter(isError) })).filter(fileWithMessages => fileWithMessages.messages.length > 0);
    console.error(prettyPrintMessages(rootPath, filesWithErrors));
  } else {
    const timeTaken = new Date().getTime() - start;
    console.log(`Linted in ${(timeTaken / 1000).toFixed(2)}s:`);
    console.log(`${errorCount} error(s), ${warningCount} warning(s), ${infoCount} info(s)`);
  }
  if (errorCount > 0) {
    return 1;
  } else if (warningCount > options.maxWarnings && options.maxWarnings !== -1) {
    return 1;
  } else if (infoCount > options.maxInfo && options.maxInfo !== -1) {
    return 1;
  }
  return 0;
}
