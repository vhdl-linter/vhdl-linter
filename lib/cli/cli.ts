#!/usr/bin/env node
import { cli } from './cliExec';

(async () => {
  const returnCode = await cli(process.argv);
  process.exit(returnCode);
})().catch(err => console.error(err));
