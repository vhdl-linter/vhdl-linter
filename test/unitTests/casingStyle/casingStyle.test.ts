import { expect, test } from '@jest/globals';
import { join } from 'path';
import { pathToFileURL } from 'url';
import { ProjectParser } from '../../../lib/projectParser';
import { VhdlLinter } from '../../../lib/vhdlLinter';
import { readFileSyncNorm } from "../../../lib/cli/readFileSyncNorm";
import { overwriteSettings } from '../../../lib/settingsUtil';
test.each([
  'object',
  'constantGeneric',
].flatMap(file => [
  [file, 'snake_case'],
  [file, 'PascalCase'],
  [file, 'camelCase'],
  [file, 'CONSTANT_CASE'],
  [file, 'ignore']
]))('testing casing style for %s with %s"', async (file: string, casing: 'snake_case' | 'PascalCase' | 'camelCase' | 'CONSTANT_CASE' | 'ignore') => {
  const overwrite: Record<string, string> = {};
  overwrite[`${file}Casing`] = casing;
  const path = join(__dirname, `${file}.vhd`);
  const projectParser = await ProjectParser.create([pathToFileURL(__dirname)]);
  const settings = overwriteSettings(await projectParser.getDocumentSettings(pathToFileURL(path)),{
    style: overwrite
  });
  const linter = new VhdlLinter(pathToFileURL(path), readFileSyncNorm(path, { encoding: 'utf8' }),
    projectParser, settings);
  await linter.checkAll();

  expect(linter.messages).toMatchSnapshot();
  const codes = linter.messages.filter(message => message.message.includes('(casing-style)')).map(message => String(message.code ?? '').split(String(';'))).flat();
  const solutions = (await Promise.all(codes.map(async code => await linter.diagnosticCodeActionResolveRegistry[parseInt(code)]?.(linter.uri.toString())))).flat()
    .filter(solution => solution?.title.includes('Replace with'));
  expect(solutions.map(solution => ({
    ...solution,
    edit: solution?.edit?.changes?.[pathToFileURL(path).toString()],

  }))).toMatchSnapshot();
  await projectParser.stop();
});