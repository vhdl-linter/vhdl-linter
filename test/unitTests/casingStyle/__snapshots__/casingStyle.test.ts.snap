// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing casing style for constantGeneric with CONSTANT_CASE" 1`] = `
[
  {
    "code": "4;5",
    "message": "i_snake_case does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "InPascalCase does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 3,
      },
      "start": {
        "character": 4,
        "line": 3,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "camelCase does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 22,
        "line": 7,
      },
      "start": {
        "character": 13,
        "line": 7,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for constantGeneric with CONSTANT_CASE" 2`] = `
[
  {
    "edit": [
      {
        "newText": "I_SNAKE_CASE",
        "range": {
          "end": {
            "character": 16,
            "line": 2,
          },
          "start": {
            "character": 4,
            "line": 2,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'I_SNAKE_CASE'",
  },
  {
    "edit": [
      {
        "newText": "IN_PASCAL_CASE",
        "range": {
          "end": {
            "character": 16,
            "line": 3,
          },
          "start": {
            "character": 4,
            "line": 3,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'IN_PASCAL_CASE'",
  },
  {
    "edit": [
      {
        "newText": "CAMEL_CASE",
        "range": {
          "end": {
            "character": 22,
            "line": 7,
          },
          "start": {
            "character": 13,
            "line": 7,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'CAMEL_CASE'",
  },
]
`;

exports[`testing casing style for constantGeneric with PascalCase" 1`] = `
[
  {
    "code": "4;5",
    "message": "i_snake_case does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "camelCase does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 22,
        "line": 7,
      },
      "start": {
        "character": 13,
        "line": 7,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "CONSTANT_CASE does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 26,
        "line": 8,
      },
      "start": {
        "character": 13,
        "line": 8,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for constantGeneric with PascalCase" 2`] = `
[
  {
    "edit": [
      {
        "newText": "ISnakeCase",
        "range": {
          "end": {
            "character": 16,
            "line": 2,
          },
          "start": {
            "character": 4,
            "line": 2,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'ISnakeCase'",
  },
  {
    "edit": [
      {
        "newText": "CamelCase",
        "range": {
          "end": {
            "character": 22,
            "line": 7,
          },
          "start": {
            "character": 13,
            "line": 7,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'CamelCase'",
  },
  {
    "edit": [
      {
        "newText": "ConstantCase",
        "range": {
          "end": {
            "character": 26,
            "line": 8,
          },
          "start": {
            "character": 13,
            "line": 8,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'ConstantCase'",
  },
]
`;

exports[`testing casing style for constantGeneric with camelCase" 1`] = `
[
  {
    "code": "4;5",
    "message": "i_snake_case does not match the style settings (camelCase) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "InPascalCase does not match the style settings (camelCase) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 3,
      },
      "start": {
        "character": 4,
        "line": 3,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "CONSTANT_CASE does not match the style settings (camelCase) (casing-style)",
    "range": {
      "end": {
        "character": 26,
        "line": 8,
      },
      "start": {
        "character": 13,
        "line": 8,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for constantGeneric with camelCase" 2`] = `
[
  {
    "edit": [
      {
        "newText": "iSnakeCase",
        "range": {
          "end": {
            "character": 16,
            "line": 2,
          },
          "start": {
            "character": 4,
            "line": 2,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'iSnakeCase'",
  },
  {
    "edit": [
      {
        "newText": "inPascalCase",
        "range": {
          "end": {
            "character": 16,
            "line": 3,
          },
          "start": {
            "character": 4,
            "line": 3,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'inPascalCase'",
  },
  {
    "edit": [
      {
        "newText": "constantCase",
        "range": {
          "end": {
            "character": 26,
            "line": 8,
          },
          "start": {
            "character": 13,
            "line": 8,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'constantCase'",
  },
]
`;

exports[`testing casing style for constantGeneric with ignore" 1`] = `[]`;

exports[`testing casing style for constantGeneric with ignore" 2`] = `[]`;

exports[`testing casing style for constantGeneric with snake_case" 1`] = `
[
  {
    "code": "4;5",
    "message": "InPascalCase does not match the style settings (snake_case) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 3,
      },
      "start": {
        "character": 4,
        "line": 3,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "camelCase does not match the style settings (snake_case) (casing-style)",
    "range": {
      "end": {
        "character": 22,
        "line": 7,
      },
      "start": {
        "character": 13,
        "line": 7,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "CONSTANT_CASE does not match the style settings (snake_case) (casing-style)",
    "range": {
      "end": {
        "character": 26,
        "line": 8,
      },
      "start": {
        "character": 13,
        "line": 8,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for constantGeneric with snake_case" 2`] = `
[
  {
    "edit": [
      {
        "newText": "in_pascal_case",
        "range": {
          "end": {
            "character": 16,
            "line": 3,
          },
          "start": {
            "character": 4,
            "line": 3,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'in_pascal_case'",
  },
  {
    "edit": [
      {
        "newText": "camel_case",
        "range": {
          "end": {
            "character": 22,
            "line": 7,
          },
          "start": {
            "character": 13,
            "line": 7,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'camel_case'",
  },
  {
    "edit": [
      {
        "newText": "constant_case",
        "range": {
          "end": {
            "character": 26,
            "line": 8,
          },
          "start": {
            "character": 13,
            "line": 8,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'constant_case'",
  },
]
`;

exports[`testing casing style for object with CONSTANT_CASE" 1`] = `
[
  {
    "code": "4;5",
    "message": "ent does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 10,
        "line": 0,
      },
      "start": {
        "character": 7,
        "line": 0,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "i_snake_case does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "InPascalCase does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 3,
      },
      "start": {
        "character": 4,
        "line": 3,
      },
    },
    "severity": 3,
  },
  {
    "code": "10;11",
    "message": "arch does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 17,
        "line": 6,
      },
      "start": {
        "character": 13,
        "line": 6,
      },
    },
    "severity": 3,
  },
  {
    "code": "12;13",
    "message": "camelCase does not match the style settings (CONSTANT_CASE) (casing-style)",
    "range": {
      "end": {
        "character": 20,
        "line": 7,
      },
      "start": {
        "character": 11,
        "line": 7,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for object with CONSTANT_CASE" 2`] = `
[
  {
    "edit": [
      {
        "newText": "ENT",
        "range": {
          "end": {
            "character": 10,
            "line": 0,
          },
          "start": {
            "character": 7,
            "line": 0,
          },
        },
      },
      {
        "newText": "ENT",
        "range": {
          "end": {
            "character": 24,
            "line": 6,
          },
          "start": {
            "character": 21,
            "line": 6,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'ENT'",
  },
  {
    "edit": [
      {
        "newText": "I_SNAKE_CASE",
        "range": {
          "end": {
            "character": 16,
            "line": 2,
          },
          "start": {
            "character": 4,
            "line": 2,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'I_SNAKE_CASE'",
  },
  {
    "edit": [
      {
        "newText": "IN_PASCAL_CASE",
        "range": {
          "end": {
            "character": 16,
            "line": 3,
          },
          "start": {
            "character": 4,
            "line": 3,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'IN_PASCAL_CASE'",
  },
  {
    "edit": [
      {
        "newText": "ARCH",
        "range": {
          "end": {
            "character": 17,
            "line": 6,
          },
          "start": {
            "character": 13,
            "line": 6,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'ARCH'",
  },
  {
    "edit": [
      {
        "newText": "CAMEL_CASE",
        "range": {
          "end": {
            "character": 20,
            "line": 7,
          },
          "start": {
            "character": 11,
            "line": 7,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'CAMEL_CASE'",
  },
]
`;

exports[`testing casing style for object with PascalCase" 1`] = `
[
  {
    "code": "4;5",
    "message": "ent does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 10,
        "line": 0,
      },
      "start": {
        "character": 7,
        "line": 0,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "i_snake_case does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "arch does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 17,
        "line": 6,
      },
      "start": {
        "character": 13,
        "line": 6,
      },
    },
    "severity": 3,
  },
  {
    "code": "10;11",
    "message": "camelCase does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 20,
        "line": 7,
      },
      "start": {
        "character": 11,
        "line": 7,
      },
    },
    "severity": 3,
  },
  {
    "code": "12;13",
    "message": "CONSTANT_CASE does not match the style settings (PascalCase) (casing-style)",
    "range": {
      "end": {
        "character": 24,
        "line": 8,
      },
      "start": {
        "character": 11,
        "line": 8,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for object with PascalCase" 2`] = `
[
  {
    "edit": [
      {
        "newText": "Ent",
        "range": {
          "end": {
            "character": 10,
            "line": 0,
          },
          "start": {
            "character": 7,
            "line": 0,
          },
        },
      },
      {
        "newText": "Ent",
        "range": {
          "end": {
            "character": 24,
            "line": 6,
          },
          "start": {
            "character": 21,
            "line": 6,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'Ent'",
  },
  {
    "edit": [
      {
        "newText": "ISnakeCase",
        "range": {
          "end": {
            "character": 16,
            "line": 2,
          },
          "start": {
            "character": 4,
            "line": 2,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'ISnakeCase'",
  },
  {
    "edit": [
      {
        "newText": "Arch",
        "range": {
          "end": {
            "character": 17,
            "line": 6,
          },
          "start": {
            "character": 13,
            "line": 6,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'Arch'",
  },
  {
    "edit": [
      {
        "newText": "CamelCase",
        "range": {
          "end": {
            "character": 20,
            "line": 7,
          },
          "start": {
            "character": 11,
            "line": 7,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'CamelCase'",
  },
  {
    "edit": [
      {
        "newText": "ConstantCase",
        "range": {
          "end": {
            "character": 24,
            "line": 8,
          },
          "start": {
            "character": 11,
            "line": 8,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'ConstantCase'",
  },
]
`;

exports[`testing casing style for object with camelCase" 1`] = `
[
  {
    "code": "4;5",
    "message": "i_snake_case does not match the style settings (camelCase) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 2,
      },
      "start": {
        "character": 4,
        "line": 2,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "InPascalCase does not match the style settings (camelCase) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 3,
      },
      "start": {
        "character": 4,
        "line": 3,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "CONSTANT_CASE does not match the style settings (camelCase) (casing-style)",
    "range": {
      "end": {
        "character": 24,
        "line": 8,
      },
      "start": {
        "character": 11,
        "line": 8,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for object with camelCase" 2`] = `
[
  {
    "edit": [
      {
        "newText": "iSnakeCase",
        "range": {
          "end": {
            "character": 16,
            "line": 2,
          },
          "start": {
            "character": 4,
            "line": 2,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'iSnakeCase'",
  },
  {
    "edit": [
      {
        "newText": "inPascalCase",
        "range": {
          "end": {
            "character": 16,
            "line": 3,
          },
          "start": {
            "character": 4,
            "line": 3,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'inPascalCase'",
  },
  {
    "edit": [
      {
        "newText": "constantCase",
        "range": {
          "end": {
            "character": 24,
            "line": 8,
          },
          "start": {
            "character": 11,
            "line": 8,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'constantCase'",
  },
]
`;

exports[`testing casing style for object with ignore" 1`] = `[]`;

exports[`testing casing style for object with ignore" 2`] = `[]`;

exports[`testing casing style for object with snake_case" 1`] = `
[
  {
    "code": "4;5",
    "message": "InPascalCase does not match the style settings (snake_case) (casing-style)",
    "range": {
      "end": {
        "character": 16,
        "line": 3,
      },
      "start": {
        "character": 4,
        "line": 3,
      },
    },
    "severity": 3,
  },
  {
    "code": "6;7",
    "message": "camelCase does not match the style settings (snake_case) (casing-style)",
    "range": {
      "end": {
        "character": 20,
        "line": 7,
      },
      "start": {
        "character": 11,
        "line": 7,
      },
    },
    "severity": 3,
  },
  {
    "code": "8;9",
    "message": "CONSTANT_CASE does not match the style settings (snake_case) (casing-style)",
    "range": {
      "end": {
        "character": 24,
        "line": 8,
      },
      "start": {
        "character": 11,
        "line": 8,
      },
    },
    "severity": 3,
  },
]
`;

exports[`testing casing style for object with snake_case" 2`] = `
[
  {
    "edit": [
      {
        "newText": "in_pascal_case",
        "range": {
          "end": {
            "character": 16,
            "line": 3,
          },
          "start": {
            "character": 4,
            "line": 3,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'in_pascal_case'",
  },
  {
    "edit": [
      {
        "newText": "camel_case",
        "range": {
          "end": {
            "character": 20,
            "line": 7,
          },
          "start": {
            "character": 11,
            "line": 7,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'camel_case'",
  },
  {
    "edit": [
      {
        "newText": "constant_case",
        "range": {
          "end": {
            "character": 24,
            "line": 8,
          },
          "start": {
            "character": 11,
            "line": 8,
          },
        },
      },
    ],
    "kind": "quickfix",
    "title": "Replace with 'constant_case'",
  },
]
`;
